doc/developers-guide/Basic-iq-handler.md

```diff
@@ -0,0 +1,179 @@
+## Basic IQ Handler
+This tutorial will show you how to add & test a simple module with an IQ
+handler to MongooseIM.
+`gen_iq_handler` module provides functionality for registering IQ
+handlers for specific namespaces.
+
+
+### Clone & build
+
+See [How-to-build](../../user-guide/How-to-build) for details on building MongooseIM
+from source code.
+
+
+### Create module & add basic IQ handler
+
+* Go to `src` and create a basic module implementing `gen_mod` behaviour
+* In `start/2` register the IQ handler with specified namespace, type
+  (processing iq policy) and function which will handle the incoming IQ stanza
+* In `stop/1` remove the registered handler
+* Implement the function for handler:
+    * If the incoming IQ stanza is of type `get` or `set` it will be
+      returned with the type set to `result`
+    * If the server doesn't recognise the hostname, the returning stanza
+      will be of type `error`
+
+See [Server Rules for Processing XML Stanzas](https://tools.ietf.org/html/rfc6120#section-10) for more
+detailed information on the topic.
+
+```erlang
+-module(mod_iq_example).
+-behaviour(gen_mod).
+
+-include("mongoose.hrl").
+-include("jlib.hrl").
+
+%% gen_mod callbacks
+-export([start/2, stop/1]).
+
+%% IQ handlers
+-export([process_iq/4]).
+
+start(Host, _Opts) ->
+    gen_iq_handler:add_iq_handler(ejabberd_sm, Host, <<"erlang-solutions.com:example">>,
+                                  ?MODULE, process_iq, no_queue).
+stop(Host) ->
+    gen_iq_handler:remove_iq_handler(ejabberd_sm, Host, <<"erlang-solutions.com:example">>).
+
+process_iq(_From, _To, Acc, IQ) ->
+    IQRes = IQ#iq{type = result},
+    ?INFO_MSG("Request: ~p~n"
+              "Response: ~p~n",
+              [IQ, IQRes]),
+    {Acc, IQRes}.
+```
+
+
+### Test your handler
+
+* Go to `big_tests/tests` and create a test suite for your handler
+* Implement the test case for success and failure:
+    * In `init_per_group` register one user with a valid hostname and one
+      with invalid
+    * In each test case send IQ stanza and wait for the response
+
+```erlang
+-module(mod_iq_example_SUITE).
+-compile(export_all).
+
+-include_lib("exml/include/exml.hrl").
+
+-define(USERS, [alice, alice_bis]).
+-define(EXAMPLE_NS, <<"erlang-solutions.com:example">>).
+
+-import(distributed_helper, [mim/0,
+                             require_rpc_nodes/1,
+                             rpc/4]).
+
+%%--------------------------------------------------------------------
+%% Suite configuration
+%%--------------------------------------------------------------------
+
+all() ->
+    [{group, mod_iq_example}].
+
+groups() ->
+    G = [{mod_iq_example, [], [should_return_result,
+                               should_return_error]}],
+    ct_helper:repeat_all_until_all_ok(G).
+
+suite() ->
+    require_rpc_nodes([mim]) ++ escalus:suite().
+
+users() ->
```

Comment:
For a newbie I thought it would be more straightforward. So maybe better idea would be to just use those from `test.config` and leave a note about that?

Reply:
I think so. It's not that difficult to reuse predefined users and it's good to teach proper habits from very beginning. :)

Reply:
> it's good to teach proper habits from very beginning

Indeed, but using users predefined in `test.config` is hardly a proper habit. We're long past the point when the definitions there were truly reusable - nowadays, it's more often than not that we define new users for a new suite testing some specific functionality.

What I'm saying is that **user definitions should be added to test configuration in the suite.** With exactly the feature flags and options enabled which ought to be tested in the suite.

User definitions in `test.config` lead to abominations like `https://github.com/esl/escalus/blob/master/src/escalus_users.erl#L212-L219`, because the library assumes `escalus_users` will be in `Config` that's passed around, but since they're not added to `Config` in the suite, while the suite needs to modify them - sometimes in the scope of a test. On the other hand, if suites do not reuse users from `test.config`, but define their own, why put them into `test.config` at all? This violates data locality.